mypy errors do not fail pytest build when using ignore_missing_imports = true

When `mypy` fails with an error, `pytest --mypy` still succeeds, when setup.cfg contains `ignore_missing_imports = true`. Should fail.

When setup.cfg does not contain `ignore_missing_imports = true`, process fails for me (as it should).

```
$ py.test --mypy -m mypy 
=================================================================== test session starts ===================================================================
platform darwin -- Python 3.7.3, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /Users/tkruse/work/myapp, inifile: setup.cfg, testpaths: test
plugins: pylint-0.14.1, mypy-0.4.2, cov-2.8.1
Running mypy on 6 files... done with status 1
...
error: Argument 1... has incompatible type "str"; expected "List[str]"
...
Found 2 errors in 1 file (checked 6 source files)
=========== 6 passed, 20 deselected, 4 warnings in 1.31s

$ echo $?
0
```

mypy output:
```
$ mypy myapp --ignore-missing-imports
...
Found 2 errors in 1 file (checked 13 source files)

$ echo $?
1
```

Please see branch of this project for a small example: https://gitlab.com/tkruse/python_dummy/-/tree/pytest-mypy-bug

I produce buggy behavior like this:
```
# in virtualenv
$ python --version
Python 3.6.8

$ pip install -r setup-requirements.txt
Collecting pytest-runner
  Using cached pytest_runner-5.2-py2.py3-none-any.whl (6.8 kB)
Collecting pytest-mypy
  Downloading pytest_mypy-0.5.0-py3-none-any.whl (5.7 kB)
Collecting filelock>=3.0
Collecting pytest>=3.5; python_version >= "3.5"
Collecting mypy>=0.500; python_version >= "3.5" and python_version < "3.8"

$ py.test 
=== test session starts 
platform linux -- Python 3.6.8, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
plugins: mypy-0.5.0
...
=== mypy 

dummy/cli/dummy_main.py:11: error: Argument "description" to "ArgumentParser" has incompatible type "Optional[int]"; expected "Optional[str]"
Found 1 error in 1 file (checked 4 source files)
=== 7 passed in 0.40s 

$ echo $?
0


$ mypy dummy
dummy/cli/dummy_main.py:11: error: Argument "description" to "ArgumentParser" has incompatible type "Optional[int]"; expected "Optional[str]"
Found 1 error in 1 file (checked 4 source files)

$ echo $?
1
```

py.test gets `--mypy` from setup.cfg.

Note that I reproduced differently this time in a private project, and I did not include `ignore_missing_imports` because this already shows the problem, I hope.

I got the same with python 3.7. The example does not show the influence of `ignore_missing_imports`, because in that sample project I could reproduce the problem in a simpler way.

--- a/src/pytest_mypy.py
+++ b/src/pytest_mypy.py
@@ -73,14 +73,14 @@ def pytest_configure(config):
 
 
 def pytest_collect_file(path, parent):
-    """Create a MypyItem for every file mypy should run on."""
+    """Create a MypyFileItem for every file mypy should run on."""
     if path.ext == '.py' and any([
             parent.config.option.mypy,
             parent.config.option.mypy_ignore_missing_imports,
     ]):
-        item = MypyItem(path, parent)
+        item = MypyFileItem(path, parent)
         if nodeid_name:
-            item = MypyItem(
+            item = MypyFileItem(
                 path,
                 parent,
                 nodeid='::'.join([item.nodeid, nodeid_name]),
@@ -89,15 +89,29 @@ def pytest_collect_file(path, parent):
     return None
 
 
-class MypyItem(pytest.Item, pytest.File):
+class MypyItem(pytest.Item):
 
-    """A File that Mypy Runs On."""
+    """A Mypy-related test Item."""
 
     MARKER = 'mypy'
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.add_marker(self.MARKER)
+
+    def repr_failure(self, excinfo):
+        """
+        Unwrap mypy errors so we get a clean error message without the
+        full exception repr.
+        """
+        if excinfo.errisinstance(MypyError):
+            return excinfo.value.args[0]
+        return super().repr_failure(excinfo)
+
+
+class MypyFileItem(MypyItem, pytest.File):
+
+    """A File that Mypy Runs On."""
 
     def runtest(self):
         """Raise an exception if mypy found errors for this item."""
@@ -112,7 +126,7 @@ class MypyItem(pytest.Item, pytest.File):
                     abspaths=[
                         os.path.abspath(str(item.fspath))
                         for item in self.session.items
-                        if isinstance(item, MypyItem)
+                        if isinstance(item, MypyFileItem)
                     ],
                 )
         )
@@ -128,15 +142,6 @@ class MypyItem(pytest.Item, pytest.File):
             None,
             self.config.invocation_dir.bestrelpath(self.fspath),
         )
-
-    def repr_failure(self, excinfo):
-        """
-        Unwrap mypy errors so we get a clean error message without the
-        full exception repr.
-        """
-        if excinfo.errisinstance(MypyError):
-            return excinfo.value.args[0]
-        return super().repr_failure(excinfo)
 
 
 def _cached_json_results(results_path, results_factory=None):

--- a/src/pytest_mypy.py
+++ b/src/pytest_mypy.py
@@ -1,5 +1,6 @@
 """Mypy static type checker plugin for Pytest"""
 
+import functools
 import json
 import os
 from tempfile import NamedTemporaryFile
@@ -115,21 +116,7 @@ class MypyFileItem(MypyItem, pytest.File):
 
     def runtest(self):
         """Raise an exception if mypy found errors for this item."""
-        results = _cached_json_results(
-            results_path=(
-                self.config._mypy_results_path
-                if _is_master(self.config) else
-                self.config.slaveinput['_mypy_results_path']
-            ),
-            results_factory=lambda:
-                _mypy_results_factory(
-                    abspaths=[
-                        os.path.abspath(str(item.fspath))
-                        for item in self.session.items
-                        if isinstance(item, MypyFileItem)
-                    ],
-                )
-        )
+        results = _mypy_results(self.session)
         abspath = os.path.abspath(str(self.fspath))
         errors = results['abspath_errors'].get(abspath)
         if errors:
@@ -142,6 +129,25 @@ class MypyFileItem(MypyItem, pytest.File):
             None,
             self.config.invocation_dir.bestrelpath(self.fspath),
         )
+
+
+def _mypy_results(session):
+    """Get the cached mypy results for the session, or generate them."""
+    return _cached_json_results(
+        results_path=(
+            session.config._mypy_results_path
+            if _is_master(session.config) else
+            session.config.slaveinput['_mypy_results_path']
+        ),
+        results_factory=functools.partial(
+            _mypy_results_factory,
+            abspaths=[
+                os.path.abspath(str(item.fspath))
+                for item in session.items
+                if isinstance(item, MypyFileItem)
+            ],
+        )
+    )
 
 
 def _cached_json_results(results_path, results_factory=None):

--- a/src/pytest_mypy.py
+++ b/src/pytest_mypy.py
@@ -90,6 +90,24 @@ def pytest_collect_file(path, parent):
     return None
 
 
+@pytest.hookimpl(hookwrapper=True, trylast=True)
+def pytest_collection_modifyitems(session, config, items):
+    """
+    Add a MypyStatusItem if any MypyFileItems were collected.
+
+    Since mypy might check files that were not collected,
+    pytest could pass even though mypy failed!
+    To prevent that, add an explicit check for the mypy exit status.
+
+    This should execute as late as possible to avoid missing any
+    MypyFileItems injected by other pytest_collection_modifyitems
+    implementations.
+    """
+    yield
+    if any(isinstance(item, MypyFileItem) for item in items):
+        items.append(MypyStatusItem(nodeid_name, session, config, session))
+
+
 class MypyItem(pytest.Item):
 
     """A Mypy-related test Item."""
@@ -129,6 +147,21 @@ class MypyFileItem(MypyItem, pytest.File):
             None,
             self.config.invocation_dir.bestrelpath(self.fspath),
         )
+
+
+class MypyStatusItem(MypyItem):
+
+    """A check for a non-zero mypy exit status."""
+
+    def runtest(self):
+        """Raise a MypyError if mypy exited with a non-zero status."""
+        results = _mypy_results(self.session)
+        if results['status']:
+            raise MypyError(
+                'mypy exited with status {status}.'.format(
+                    status=results['status'],
+                ),
+            )
 
 
 def _mypy_results(session):

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -14,29 +14,29 @@ def xdist_args(request):
     return ['-n', 'auto'] if request.param else []
 
 
-@pytest.mark.parametrize('test_count', [1, 2])
-def test_mypy_success(testdir, test_count, xdist_args):
+@pytest.mark.parametrize('pyfile_count', [1, 2])
+def test_mypy_success(testdir, pyfile_count, xdist_args):
     """Verify that running on a module with no type errors passes."""
     testdir.makepyfile(
         **{
-            'test_' + str(test_i): '''
-                def myfunc(x: int) -> int:
+            'pyfile_' + str(pyfile_i): '''
+                def pyfunc(x: int) -> int:
                     return x * 2
             '''
-            for test_i in range(test_count)
+            for pyfile_i in range(pyfile_count)
         }
     )
     result = testdir.runpytest_subprocess(*xdist_args)
     result.assert_outcomes()
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(passed=test_count)
+    result.assert_outcomes(passed=pyfile_count)
     assert result.ret == 0
 
 
 def test_mypy_error(testdir, xdist_args):
     """Verify that running on a module with type errors fails."""
     testdir.makepyfile('''
-        def myfunc(x: int) -> str:
+        def pyfunc(x: int) -> str:
             return x * 2
     ''')
     result = testdir.runpytest_subprocess(*xdist_args)
@@ -178,7 +178,7 @@ def test_pytest_collection_modifyitems(testdir, xdist_args):
                 items.pop(mypy_item_i)
     ''')
     testdir.makepyfile('''
-        def myfunc(x: int) -> str:
+        def pyfunc(x: int) -> str:
             return x * 2
 
         def test_pass():

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -167,6 +167,11 @@ def test_api_nodeid_name(testdir, xdist_args):
 
 
 def test_pytest_collection_modifyitems(testdir, xdist_args):
+    """
+    Verify that collected files which are removed in a
+    pytest_collection_modifyitems implementation are not
+    checked by mypy.
+    """
     testdir.makepyfile(conftest='''
         def pytest_collection_modifyitems(session, config, items):
             plugin = config.pluginmanager.getplugin('mypy')

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -54,13 +54,19 @@ def test_mypy_ignore_missings_imports(testdir, xdist_args):
     Verify that --mypy-ignore-missing-imports
     causes mypy to ignore missing imports.
     """
+    module_name = 'is_always_missing'
     testdir.makepyfile('''
-        import pytest_mypy
-    ''')
+        try:
+            import {module_name}
+        except ImportError:
+            pass
+    '''.format(module_name=module_name))
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
     result.assert_outcomes(failed=1)
     result.stdout.fnmatch_lines([
-        "1: error: Cannot find *module named 'pytest_mypy'",
+        "2: error: Cannot find *module named '{module_name}'".format(
+            module_name=module_name,
+        ),
     ])
     assert result.ret != 0
     result = testdir.runpytest_subprocess(
@@ -88,13 +94,15 @@ def test_mypy_marker(testdir, xdist_args):
 def test_non_mypy_error(testdir, xdist_args):
     """Verify that non-MypyError exceptions are passed through the plugin."""
     message = 'This is not a MypyError.'
-    testdir.makepyfile('''
-        import pytest_mypy
+    testdir.makepyfile(conftest='''
+        def pytest_configure(config):
+            plugin = config.pluginmanager.getplugin('mypy')
 
-        def _patched_runtest(*args, **kwargs):
-            raise Exception('{message}')
+            class PatchedMypyItem(plugin.MypyItem):
+                def runtest(self):
+                    raise Exception('{message}')
 
-        pytest_mypy.MypyItem.runtest = _patched_runtest
+            plugin.MypyItem = PatchedMypyItem
     '''.format(message=message))
     result = testdir.runpytest_subprocess(*xdist_args)
     result.assert_outcomes()

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -29,7 +29,10 @@ def test_mypy_success(testdir, pyfile_count, xdist_args):
     result = testdir.runpytest_subprocess(*xdist_args)
     result.assert_outcomes()
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(passed=pyfile_count)
+    mypy_file_checks = pyfile_count
+    mypy_status_check = 1
+    mypy_checks = mypy_file_checks + mypy_status_check
+    result.assert_outcomes(passed=mypy_checks)
     assert result.ret == 0
 
 
@@ -42,7 +45,10 @@ def test_mypy_error(testdir, xdist_args):
     result = testdir.runpytest_subprocess(*xdist_args)
     result.assert_outcomes()
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(failed=1)
+    mypy_file_checks = 1
+    mypy_status_check = 1
+    mypy_checks = mypy_file_checks + mypy_status_check
+    result.assert_outcomes(failed=mypy_checks)
     result.stdout.fnmatch_lines([
         '2: error: Incompatible return value*',
     ])
@@ -62,7 +68,10 @@ def test_mypy_ignore_missings_imports(testdir, xdist_args):
             pass
     '''.format(module_name=module_name))
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(failed=1)
+    mypy_file_checks = 1
+    mypy_status_check = 1
+    mypy_checks = mypy_file_checks + mypy_status_check
+    result.assert_outcomes(failed=mypy_checks)
     result.stdout.fnmatch_lines([
         "2: error: Cannot find *module named '{module_name}'".format(
             module_name=module_name,
@@ -73,7 +82,7 @@ def test_mypy_ignore_missings_imports(testdir, xdist_args):
         '--mypy-ignore-missing-imports',
         *xdist_args
     )
-    result.assert_outcomes(passed=1)
+    result.assert_outcomes(passed=mypy_checks)
     assert result.ret == 0
 
 
@@ -84,10 +93,14 @@ def test_mypy_marker(testdir, xdist_args):
             assert False
     ''')
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(failed=1, passed=1)
+    test_count = 1
+    mypy_file_checks = 1
+    mypy_status_check = 1
+    mypy_checks = mypy_file_checks + mypy_status_check
+    result.assert_outcomes(failed=test_count, passed=mypy_checks)
     assert result.ret != 0
     result = testdir.runpytest_subprocess('--mypy', '-m', 'mypy', *xdist_args)
-    result.assert_outcomes(passed=1)
+    result.assert_outcomes(passed=mypy_checks)
     assert result.ret == 0
 
 
@@ -107,7 +120,12 @@ def test_non_mypy_error(testdir, xdist_args):
     result = testdir.runpytest_subprocess(*xdist_args)
     result.assert_outcomes()
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(failed=1)
+    mypy_file_checks = 1  # conftest.py
+    mypy_status_check = 1
+    result.assert_outcomes(
+        failed=mypy_file_checks,  # patched to raise an Exception
+        passed=mypy_status_check,  # conftest.py has no type errors.
+    )
     result.stdout.fnmatch_lines(['*' + message])
     assert result.ret != 0
 
@@ -171,6 +189,9 @@ def test_pytest_collection_modifyitems(testdir, xdist_args):
     Verify that collected files which are removed in a
     pytest_collection_modifyitems implementation are not
     checked by mypy.
+
+    This would also fail if a MypyStatusItem were injected
+    despite there being no MypyFileItems.
     """
     testdir.makepyfile(conftest='''
         def pytest_collection_modifyitems(session, config, items):
@@ -190,5 +211,49 @@ def test_pytest_collection_modifyitems(testdir, xdist_args):
             pass
     ''')
     result = testdir.runpytest_subprocess('--mypy', *xdist_args)
-    result.assert_outcomes(passed=1)
+    test_count = 1
+    result.assert_outcomes(passed=test_count)
     assert result.ret == 0
+
+
+def test_mypy_indirect(testdir, xdist_args):
+    """Verify that uncollected files checked by mypy cause a failure."""
+    testdir.makepyfile(bad='''
+        def pyfunc(x: int) -> str:
+            return x * 2
+    ''')
+    testdir.makepyfile(good='''
+        import bad
+    ''')
+    xdist_args.append('good.py')  # Nothing may come after xdist_args in py34.
+    result = testdir.runpytest_subprocess('--mypy', *xdist_args)
+    assert result.ret != 0
+
+
+def test_mypy_indirect_inject(testdir, xdist_args):
+    """
+    Verify that uncollected files checked by mypy because of a MypyFileItem
+    injected in pytest_collection_modifyitems cause a failure.
+    """
+    testdir.makepyfile(bad='''
+        def pyfunc(x: int) -> str:
+            return x * 2
+    ''')
+    testdir.makepyfile(good='''
+        import bad
+    ''')
+    testdir.makepyfile(conftest='''
+        import py
+        import pytest
+
+        @pytest.hookimpl(trylast=True)  # Inject as late as possible.
+        def pytest_collection_modifyitems(session, config, items):
+            plugin = config.pluginmanager.getplugin('mypy')
+            items.append(
+                plugin.MypyFileItem(py.path.local('good.py'), session),
+            )
+    ''')
+    testdir.mkdir('empty')
+    xdist_args.append('empty')  # Nothing may come after xdist_args in py34.
+    result = testdir.runpytest_subprocess('--mypy', *xdist_args)
+    assert result.ret != 0

--- a/tests/test_pytest_mypy.py
+++ b/tests/test_pytest_mypy.py
@@ -98,11 +98,11 @@ def test_non_mypy_error(testdir, xdist_args):
         def pytest_configure(config):
             plugin = config.pluginmanager.getplugin('mypy')
 
-            class PatchedMypyItem(plugin.MypyItem):
+            class PatchedMypyFileItem(plugin.MypyFileItem):
                 def runtest(self):
                     raise Exception('{message}')
 
-            plugin.MypyItem = PatchedMypyItem
+            plugin.MypyFileItem = PatchedMypyFileItem
     '''.format(message=message))
     result = testdir.runpytest_subprocess(*xdist_args)
     result.assert_outcomes()
@@ -178,7 +178,7 @@ def test_pytest_collection_modifyitems(testdir, xdist_args):
             for mypy_item_i in reversed([
                     i
                     for i, item in enumerate(items)
-                    if isinstance(item, plugin.MypyItem)
+                    if isinstance(item, plugin.MypyFileItem)
             ]):
                 items.pop(mypy_item_i)
     ''')

--- a/tox.ini
+++ b/tox.ini
@@ -80,7 +80,7 @@ deps =
 
     pytest-cov ~= 2.5.1
     pytest-randomly ~= 2.1.1
-commands = py.test -p no:mypy -n auto --cov pytest_mypy --cov-fail-under 100 --cov-report term-missing {posargs}
+commands = py.test -p no:mypy --cov pytest_mypy --cov-fail-under 100 --cov-report term-missing {posargs:-n auto} tests
 
 [testenv:static]
 deps =

